ADBDEV-3852 Fix SUBLINK processing during query normalization in ORCA

bimboterminator1 opened the pull request:
Fix ORCA SubLink processing at query normalization
The ORCA optimizer could produce an invalid query plan or fallback to the legacy
optimizer due to incorrect processing of nested SubLinks (SubLink contains one
more SubLink or rtable subquery inside) during query normalization.

Firstly, the incorrect behaviour took place when a query had a nested
SubLink in its targetList, and SubLink lower levels contained attributes that
were referenced in the query's GROUP BY clause. ORCA could construct for
such queries an invalid plan, according to which some wrong and unexpected
attributes got into the nested part of SubLink's plan (the example is provided
in tests section). And the presence of those attributes could lead to incorrect
query execution or even segmentation fault.

Inside the NormalizeGroupByProjList function the context.m_lower_table_tlist,
which contains grouping attributes from target_list_copy, is created. And the
entries of target_list_copy, that are not grouping attributes, are mutated
inside the RunExtractAggregatesMutator in order to make their inner expressions
reference the corresponding entries from context.m_lower_table_tlist.
The invalid plan could be produced because of incorrect mapping between Var's
varattno (in target_list_copy) and some TargetEntry's resno
(in context.m_lower_table_tlist). This mapping is established by
RunExtractAggregatesMutator, where for the Var code branch the following
assignment takes place: var->varattno = found_tle->resno, where found_tle is the
corresponding entry from context.m_lower_table_tlist. However, the behaviour of
this function was erroneous for the case of nested SubLinks because it
mutated only the upper level of nested SubLink structure, and other subqueries
remained unmodified because RunExtractAggregatesMutator did not have a branch
for the Query, which was required for futher diving into lower subqueries. And
because inner Vars hadn't been mutated, they started reference wrong attributes.

This patch modifies RunExtractAggregatesMutator function by adding the Query
branch to make the mutator correctly step into nested SubLink queries. And the
call of MutateQueryOrExpressionTree is replaced with the call of mutator itself
in order to correctly modify m_current_query_level when stepping into lower
subqueries.

Secondly, the erroneous behaviour took place when nested SubLink was in
GROUP BY clause itself (see the test section). In this case nested SubLinks were
mutated incorrectly by RunGroupingColMutator.
RunGroupingColMutator had Query branch, however, when mutating the subquery
of the SubLink (at second query level), the m_current_query_level variable was
not increased respectively, because MutateQueryOrExpressionTree, which was
responsible for calling the RunGroupingColMutator on the lower subquery, did not
know anything about a context. Therefore, for the query like in the tests, wrong
value of m_current_query_level led to unnecessary modification of the Var's
varlevelsup value. And when Var's varlevelsup value had been unnecessary
modified, it started referencing different query level. This could lead to the
fallback or invalid plan construction at further planning stages.

This patch fixes erroneous behaviour of RunGroupingColMutator by replacing the
call of MutateQueryOrExpressionTree with the call of RunGroupingColMutator in
order to properly modify m_current_query_level when mutating lower level rtes.

Thirdly, similar issue could arise when query had correlated subquery in SubLink
under the AggRef.
While mutating the SubLink inside RunExtractAggregatesMutator, the Aggref branch
is executed for SubLink's query targetList, which mutates args of the AggRef.
And if a Var had varlevelsup value greater than the agglevelsup, i.e the Var
referenced a relation that is higher than the AggRef, the Var was not mutated
at all. The varlevelsup field was not modified and because of that, when AggRef
was pulled up to the zero query level (by appending it to the
context.m_lower_table_tlist), the Var, whose varlevelsup was higher than
AggRef's, and which remained unchanged, started referencing the wrong relation.
This could lead to the fallback or invalid plan construction when processing
attributes at further planning stages.

This patch makes the RunExtractAggregatesMutator modify the varlevelsup of the
Var relatively the AggRef level in order to preserve proper query level.

bimboterminator1 committed:
Fix ORCA SubLink processing at query normalization

The ORCA optimizer may produce an invalid query plan or fallback to the legacy
optimizer due to incorrect processing of nested SubLink entries during query
normalization.
Firstly, the issue occurs when a query has a multi-level SubLink entry in its
targetList, and this SubLink contains attributes that are referenced in the
query's GROUP BY clause. Currently, ORCA may construct for such queries an
invalid plan, according to which some wrong and unexpected attributes get into
the nested part of SubLink's plan (example of such query is provided in
tests section). And the presence of those attributed may lead to incorrect query
execution or even segmentation fault.

The ORCA during its work constructs a DXL representation of the Query tree, and
in order to make the final tree as simple as possible it normalizes original
Query structures via CQueryMutators::NormalizeQuery function. Inside the
NormalizeQuery function the NormalizeGroupByProjList function is called, where
the problems, that are considered in this patch, occur. In details, inside
NormalizeGroupByProjList the ConvertToDerivedTable is called, which returns a
new query that wraps the original query as rtable entry. Then the original
(derived) query's targetList is being iterated, and for entries, which are
grouping columns, FixGroupingCols function is called, where each targetEntry
is being mutated with RunGroupingColMutator mutator function. Then this entry
is added to a mutator context's (SContextGrpbyPlMutator) targetList
(context.m_lower_table_tlist) and also replaces the Expr for the entry in
original targetList to make it reference the newly added entry in the
m_lower_table_list. Therefore, because not all entries of original targetList
are grouping columns and not all are being mutated, the
context.m_lower_table_tlist may become shorter than original. After iteration
process, the entries of original targetList, that are not grouping columns,
are modified by RunExtractAggregatesMutator function, which checks whether the
entry contains attributes from grouping clause and replaces them with
corresponding expression from context.m_lower_table_tlist. Here, during
replacement, the original expression's varattno is assigned to
context.m_lower_table_tlist entry resno, also the varlevelsup is also being
changed corresponding to the current query level. And if these values will be
incorrect or remain unmodified, the ORCA may built invalid plan or fallback to
legacy optimizer. This happens because the new context.m_lower_table_tlist is
then being saved in derived query and, as it was said before, it can have
length and order of elements different from original, and, therefore, if the
mutation process is wrong, the entries of modified original targetList may not
correspond to the right entries of new context.m_lower_table_tlist, which leads
to invalid plan construction. (there is a mechanism of hash maps, if they are
filled wrongly due to wrong targetLists, the mapping vars -> relation column
will be incorrect).

When a query contains nested SubLinks, RunExtractAggregatesMutator mutates
only the upper level of nested structure, and other subqueries remain
unmodified, because RunExtractAggregatesMutator does not have a branch for the
Query, which is required for futher diving into lower subqueries.

This patch modifies RunExtractAggregatesMutator function by adding the Query
branch to make the mutator correctly step into nested SubLink queries. And the
call of MutateQueryOrExpressionTree is replaced to the call of mutator itself
in order to correctly modify m_current_query_level when stepping into lower
subqueries.

Secondly, the ORCA may fallback to the legacy optimizer or build invalid plan,
when nested SubLink entry is in GROUP BY clause itself (see the test
section). In this case nested SubLinks are mutated incorrectly.
As it was mentioned earlier, inside the NormalizeGroupByProjList function the
derived query's targetList is being iterated, and for entries, which are
grouping columns, FixGroupingCols function is called, where each targetEntry
is being mutated with RunGroupingColMutator mutator function.
RunGroupingColMutator handles multi-level SubLink incorrectly as well. After
stepping into MutateQueryOrExpressionTree inside the SubLink branch and
mutating upper query level, the mutation on lower query rte is called,
and RunGroupingColMutator is called for this level. It has Query node type to
process, however, when diving into second level deep SubLink subquery, the
m_current_query_level variable is not increased, because
MutateQueryOrExpressionTree, which is responsible for calling mutator on the
lower subquery, does not know anything about a context. Therefore, for the
query like in the tests, that behaviour leads to extra modification of the Var's
varlevelsup value. And when Var's varlevelsup value is modified, it starts
referencing different query level. This may lead to the fallback or invalid plan
construction when accessing attributes from hash maps.

This patch fixes erroneous behaviour by replacing the call of
MutateQueryOrExpressionTree to the call of RunGroupingColMutator in order to
properly modify m_current_query_level when mutating lower level rtes.

Thirdly, the case of correlated SubLink under the AggRef is considered.
Such queries has simple structure like SELECT (SELECT max((SELECT ... FROM
s WHERE t.a=s.b))) ... FROM t. And for this case the ORCA may fallback
as well because of erroneous normalization algorithm.
In this case the NormalizeGroupByProjList is called anyway, because the query
has aggregates. While mutating the SubLink inside RunExtractAggregatesMutator,
the Aggref branch is executed for SubLink's query targetList, which mutates
args of the AggRef. And if a Var has varlevelsup value greater than the
agglevelsup, i.e the Var references a relation that is higher than the AggRef,
the Var is not mutated at all. The varlevelsup field is not modified and because
of that, when AggRef is pulled up to the derived query level (by appending it
into the context.m_lower_table_tlist inside MakeVarInDerivedTable), the Var,
whose varlevelsup was higher than AggRef's, will reference the wrong relation.
This may lead to the fallback or invalid plan construction when accessing
attributes from hash maps.

This patch makes the mutator function modify the varlevelsup of the Var
relatively the AggRef level in order to preserve proper query level.

bimboterminator1 commented:
Well, I checked the behaviour at other [types](https://github.com/arenadata/gpdb/blob/2de3b73a03045062849da9b99612bcd022334a7b/src/include/nodes/primnodes.h#L653-L662) of SubLink. They are being processed the same way, and the result is also the same. Should I add tests for other SubLink type?

red1452 reviewed (commented):
The main goal of ticket is fix SIGSEGV. Where are steps which due to segfault? Which options do you use to build cluster to get messages?
```
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  Query-to-DXL Translation: No variable entry found due to incorrect normalization of query
```
What is about the same message at `test/regress/expected/aggregate_optimizer.out`? Why did not it fix?

bimboterminator1 commented:
> The main goal of ticket is fix SIGSEGV. Where are steps which due to segfault? Which options do you use to build cluster to get messages? 

Have provided the query, which leads to SIGSEGV, in PR description. I ran simple demo cluster without mirrors. In order to display fallback messages, you should just turn on the GUC `SET optimizer_trace_fallback=on;`

> What is about the same message at `test/regress/expected/aggregate_optimizer.out`? Why did not it fix?
I will investigate this case in more details.

bimboterminator1 commented:
> What is about the same message at `test/regress/expected/aggregate_optimizer.out`? Why did not it fix?
Fixed.
Have changed the test queries. Have rewritten commit and PR messages.

bimboterminator1 commented:
@RekGRpth, I have changed the second sentence about incorrect plan, may be it will be more readable.
 In description I've provided the general description of the normalization algorithm, which current patch is fixing. If you want to get the things right, it probably  would be better to make a voice chat or eventually explore the source code

bimboterminator1 commented:
@RekGRpth, Have rewritten PR description. If its OK, I'll squash and write the commit message in similar manner.

bimboterminator1 commented:
Have squashed and rewritten the commit message.

RekGRpth commented on src/backend/gpopt/translate/CQueryMutators.cpp:
```diff
@@ -701,7 +702,7 @@ CQueryMutators::RunExtractAggregatesMutator(Node *node,
 				// Note the foo.a var which is in sum() in a subquery must now become a
 				// var referencing the current query level.
-				var->varlevelsup = 0;
+				var->varlevelsup = var->varlevelsup - context->m_agg_levels_up;
```
```c
var->varlevelsup -= context->m_agg_levels_up;
```
?

bimboterminator1 replied:
done

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3852

andr-sokolov commented on src/backend/gpopt/translate/CQueryMutators.cpp:
```diff
@@ -798,17 +799,47 @@ CQueryMutators::RunExtractAggregatesMutator(Node *node,
+
+		List *rtable = query->rtable;
+		ListCell *lc = NULL;
+		ForEach(lc, rtable)
```
I suggest to replace lines 823-825 with
```
ListCell *lc;
ForEach(lc, query->rtable)
```

lc value from 824 line is never used, because the value is rewritten in ForEach. The rtable variable is used only once, so we can use query->rtable instead of it.

bimboterminator1 replied:
Done

bimboterminator1 replied:
Wait, no, its wrong to change the pointer, the query is later returned.

bimboterminator1 replied:
Probably It is better to add a NULL check for rte
```
RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);
GPOS_ASSERT(NULL != rte);
```
Or without assert - simply return the query if rte is null

bimboterminator1 replied:
> Probably It is better to add a NULL check for rte
No, it's ok. If NULL will be inside for, the loop won't start and the query will be returned. Everything is fine.
```
#define ForEach(cell, l) \
	for ((cell) = gpdb::ListHead(l); (cell) != NULL; (cell) = lnext(cell))
```

bimboterminator1 committed:
Short code refactoring

bimboterminator1 committed:
Clang-format fix

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3852

andr-sokolov commented on src/backend/gporca/data/dxl/minidump/NestedSubLinkAndGroupBy.mdp:
```diff
@@ -0,0 +1,580 @@
+	 CREATE TABLE sublink_test (i text, j text, k numeric, m date) distributed randomly;
+
+	 explain (verbose, costs off) SELECT s.i, s.j, 1, (SELECT s.k FROM (SELECT s.k) q2) AS q1
+	 FROM sublink_test s GROUP BY s.i, s.j, q1;
```
Let's make the queries simpler:
```
CREATE TABLE t (i int, j int);
insert into t values (1, 2);

SELECT 
  j,
  (
    SELECT j
    FROM (SELECT j) q2
  )
FROM t
GROUP BY i, j;
```
The queries returned "2, 1", but not "2, 2".

andr-sokolov commented:
@bimboterminator1 , Remove please sentences about correct behaviour from the commit description. Write more about unchanged (and therefore incorrect) varattno in the description.

bimboterminator1 committed:
Test simplification

bimboterminator1 committed:
Fix ORCA SubLink processing at query normalization

The ORCA optimizer could produce an invalid query plan or fallback to the legacy
optimizer due to incorrect processing of nested SubLinks (SubLink contains one
more SubLink or rtable subquery inside) during query normalization.

Firstly, the incorrect behaviour took place when a query had a nested
SubLink in its targetList, and SubLink lower levels contained attributes that
were referenced in the query's GROUP BY clause. ORCA could construct for
such queries an invalid plan, according to which some wrong and unexpected
attributes got into the nested part of SubLink's plan (the example is provided
in tests section). And the presence of those attributes could lead to incorrect
query execution or even segmentation fault.

Inside the NormalizeGroupByProjList function the context.m_lower_table_tlist,
which contains grouping attributes from target_list_copy, is created. And the
entries of target_list_copy, that are not grouping attributes, are mutated
inside the RunExtractAggregatesMutator in order to make their inner expressions
reference the corresponding entries from context.m_lower_table_tlist.
The invalid plan could be produced because of incorrect mapping between Var's
varattno (in target_list_copy) and some TargetEntry's resno
(in context.m_lower_table_tlist). This mapping is established by
RunExtractAggregatesMutator, where for the Var code branch the following
assignment takes place: var->varattno = found_tle->resno, where found_tle is the
corresponding entry from context.m_lower_table_tlist. However, the behaviour of
this function was erroneous for the case of nested SubLinks because it
mutated only the upper level of nested SubLink structure, and other subqueries
remained unmodified because RunExtractAggregatesMutator did not have a branch
for the Query, which was required for futher diving into lower subqueries. And
because inner Vars hadn't been mutated, they started reference wrong attributes.

This patch modifies RunExtractAggregatesMutator function by adding the Query
branch to make the mutator correctly step into nested SubLink queries. And the
call of MutateQueryOrExpressionTree is replaced to the call of mutator itself
in order to correctly modify m_current_query_level when stepping into lower
subqueries.

Secondly, the erroneous behaviour took place when nested SubLink was in
GROUP BY clause itself (see the test section). In this case nested SubLinks were
mutated incorrectly by RunGroupingColMutator.
RunGroupingColMutator had Query branch, however, when mutating second level deep
SubLink subquery, the m_current_query_level variable was not increased
respectively, because MutateQueryOrExpressionTree, which was responsible for
calling the RunGroupingColMutator on the lower subquery, did not know anything
about a context. Therefore, for the query like in the tests, wrong value of
m_current_query_level led to unnecessary modification of the Var's varlevelsup
value. And when Var's varlevelsup value had been unnecessary modified, it
started referencing different query level. This could lead to the fallback or
invalid plan construction at further planning stages.

This patch fixes erroneous behaviour of RunGroupingColMutator by replacing the
call of MutateQueryOrExpressionTree to the call of RunGroupingColMutator in
order to properly modify m_current_query_level when mutating lower level rtes.

Thirdly, similar issue could arise when query had correlated SubLink under the
AggRef.
While mutating the SubLink inside RunExtractAggregatesMutator, the Aggref branch
is executed for SubLink's query targetList, which mutates args of the AggRef.
And if a Var had varlevelsup value greater than the agglevelsup, i.e the Var
referenced a relation that is higher than the AggRef, the Var was not mutated
at all. The varlevelsup field was not modified and because of that, when AggRef
was pulled up to the zero query level (by appending it to the
context.m_lower_table_tlist), the Var, whose varlevelsup was higher than
AggRef's, and which remained unchanged, started referencing the wrong relation.
This could lead to the fallback or invalid plan construction when processing
attributes at further planning stages.

This patch makes the RunExtractAggregatesMutator modify the varlevelsup of the
Var relatively the AggRef level in order to preserve proper query level.

bimboterminator1 committed:
Test description has changed

bimboterminator1 commented:
Have changed the patch description. Have simplified test cases.

RekGRpth commented on src/backend/gporca/data/dxl/minidump/NestedSubLinkAndGroupBy.mdp:
```diff
@@ -0,0 +1,410 @@
+	 case, when the SubLink column is inside the GROUP BY clause.
+	 CREATE TABLE t (i int, j int) distributed by (i);
+
+	 explain (verbose, costs off) SELECT, (SELECT j FROM (SELECT j) q2) q1
```
```sql
SELECT,
```
?

bimboterminator1 replied:
fixed

RekGRpth commented on src/backend/gporca/data/dxl/minidump/NestedSubLinkAndGroupBy.mdp:
```diff
@@ -0,0 +1,410 @@
+	 CREATE TABLE t (i int, j int) distributed by (i);
+
+	 explain (verbose, costs off) SELECT, (SELECT j FROM (SELECT j) q2) q1
+	 FROM t GROUP BY i, j q1;
```
```sql
j q1
```
?

bimboterminator1 replied:
fixed

RekGRpth commented on src/backend/gporca/data/dxl/minidump/NestedSubLinkWithGroupByAttrs.mdp:
```diff
@@ -0,0 +1,381 @@
+	 during query normalization, 
+	 CREATE TABLE t (i int, j int) distributed by (i);
+
+	 explain (verbose, costs off) SELECT, (SELECT j FROM (SELECT j) q2)
```
```sql
SELECT,
```
?

bimboterminator1 replied:
fixed

bimboterminator1 committed:
Fix ORCA SubLink processing at query normalization

The ORCA optimizer could produce an invalid query plan or fallback to the legacy
optimizer due to incorrect processing of nested SubLinks (SubLink contains one
more SubLink or rtable subquery inside) during query normalization.

Firstly, the incorrect behaviour took place when a query had a nested
SubLink in its targetList, and SubLink lower levels contained attributes that
were referenced in the query's GROUP BY clause. ORCA could construct for
such queries an invalid plan, according to which some wrong and unexpected
attributes got into the nested part of SubLink's plan (the example is provided
in tests section). And the presence of those attributes could lead to incorrect
query execution or even segmentation fault.

Inside the NormalizeGroupByProjList function the context.m_lower_table_tlist,
which contains grouping attributes from target_list_copy, is created. And the
entries of target_list_copy, that are not grouping attributes, are mutated
inside the RunExtractAggregatesMutator in order to make their inner expressions
reference the corresponding entries from context.m_lower_table_tlist.
The invalid plan could be produced because of incorrect mapping between Var's
varattno (in target_list_copy) and some TargetEntry's resno
(in context.m_lower_table_tlist). This mapping is established by
RunExtractAggregatesMutator, where for the Var code branch the following
assignment takes place: var->varattno = found_tle->resno, where found_tle is the
corresponding entry from context.m_lower_table_tlist. However, the behaviour of
this function was erroneous for the case of nested SubLinks because it
mutated only the upper level of nested SubLink structure, and other subqueries
remained unmodified because RunExtractAggregatesMutator did not have a branch
for the Query, which was required for futher diving into lower subqueries. And
because inner Vars hadn't been mutated, they started reference wrong attributes.

This patch modifies RunExtractAggregatesMutator function by adding the Query
branch to make the mutator correctly step into nested SubLink queries. And the
call of MutateQueryOrExpressionTree is replaced to the call of mutator itself
in order to correctly modify m_current_query_level when stepping into lower
subqueries.

Secondly, the erroneous behaviour took place when nested SubLink was in
GROUP BY clause itself (see the test section). In this case nested SubLinks were
mutated incorrectly by RunGroupingColMutator.
RunGroupingColMutator had Query branch, however, when mutating the subquery
of the SubLink (at second query level), the m_current_query_level variable was
not increased respectively, because MutateQueryOrExpressionTree, which was
responsible for calling the RunGroupingColMutator on the lower subquery, did not
know anything about a context. Therefore, for the query like in the tests, wrong
value of m_current_query_level led to unnecessary modification of the Var's
varlevelsup value. And when Var's varlevelsup value had been unnecessary
modified, it started referencing different query level. This could lead to the
fallback or invalid plan construction at further planning stages.

This patch fixes erroneous behaviour of RunGroupingColMutator by replacing the
call of MutateQueryOrExpressionTree to the call of RunGroupingColMutator in
order to properly modify m_current_query_level when mutating lower level rtes.

Thirdly, similar issue could arise when query had correlated subquery in SubLink
under the AggRef.
While mutating the SubLink inside RunExtractAggregatesMutator, the Aggref branch
is executed for SubLink's query targetList, which mutates args of the AggRef.
And if a Var had varlevelsup value greater than the agglevelsup, i.e the Var
referenced a relation that is higher than the AggRef, the Var was not mutated
at all. The varlevelsup field was not modified and because of that, when AggRef
was pulled up to the zero query level (by appending it to the
context.m_lower_table_tlist), the Var, whose varlevelsup was higher than
AggRef's, and which remained unchanged, started referencing the wrong relation.
This could lead to the fallback or invalid plan construction when processing
attributes at further planning stages.

This patch makes the RunExtractAggregatesMutator modify the varlevelsup of the
Var relatively the AggRef level in order to preserve proper query level.

bimboterminator1 commented:
Have changed the description a little bit for second and third cases

RekGRpth commented:
I tried testing mdp-files without the patch, but nothing broke! What are they checking then?

bimboterminator1 commented:
> I tried testing mdp-files without the patch, but nothing broke! What are they checking then?

How to write unittests otherwise? Or it's impossible to test it in thar way, because minidump has already correct Query representation?

RekGRpth commented:
Moreover, the test plan from the description of the NestedSubLinkAndGroupBy.mdp file is identical without a patch, and with a patch!

RekGRpth commented:
> How to write unittests otherwise?

Maybe filling the tables is essential?

RekGRpth commented:
Please test unit tests yourself without a patch and with a patch! If there is no difference, then you probably have to do regression tests.

bimboterminator1 commented:
The info about original query in minidump comes only in dxl form, therefore no matter are we testing with or without patch,
the minidump will generate the same plan. The unittests are not suitable for the patch because it fixes dxl query representation,
not the planner

RekGRpth commented:
> The info about original query in minidump comes only in dxl form, therefore, no matter are we testing with or without patch,
> the minidump will generate the same plan. The unittests are not suitable for the patch because it fixes dxl query representation,
> not the planner. Regression tests will be added

Then rewrite the tests for regression!

bimboterminator1 committed:
Replace mdp for regression tests

RekGRpth commented:
Comparison of the expected and obtained results for the subselect test shows that many plans have changed to orca ones. Why didn't the test fail?

RekGRpth commented on src/test/regress/sql/subselect.sql:
```diff
@@ -517,3 +517,44 @@ fetch backward all in c1;
+-- the nested SubLink are processed during query normalization for the
+-- case, when the SubLink column is inside the GROUP BY clause. The fallback
+-- shouldn't occur.
+explain (verbose, costs off)
```
This test does not check anything! Without a patch and with a patch, the plans and results are identical!

bimboterminator1 replied:
fixed, now it checks

bimboterminator1 replied:
The query didn't fallback, because all targetList entries were in group list and normalization wasn't required

bimboterminator1 committed:
Make second test work

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -517,3 +517,44 @@ fetch backward all in c1;
+-- the nested SubLink, which are referenced in GROUP BY clause, are processed
+-- during query normalization. The inner part of SubPLan should contain only t.j
+-- start_ignore
+DROP TABLE IF EXISTS t;
```
The queries should be like the nearest code - in lowercase.

bimboterminator1 replied:
fixed

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -517,3 +517,44 @@ fetch backward all in c1;
+SELECT j,
+(SELECT j FROM (SELECT j) q2)
+FROM t GROUP BY i, j;
+SELECT j,
```
Add please an empty line or make it easier to read the queries in another way

bimboterminator1 replied:
fixed

bimboterminator1 committed:
Formatting the tests

bimboterminator1 committed:
Empty line

andr-sokolov commented on src/backend/gpopt/translate/CQueryMutators.cpp:
```diff
@@ -798,17 +799,46 @@ CQueryMutators::RunExtractAggregatesMutator(Node *node,
+	{
+		Query *query = gpdb::MutateQueryTree(
+			(Query *) node,
+			(MutatorWalkerFn) CQueryMutators::RunExtractAggregatesMutator,
```
I suggest to remove redundant "CQueryMutators::"

bimboterminator1 replied:
done

andr-sokolov commented on src/backend/gpopt/translate/CQueryMutators.cpp:
```diff
@@ -798,17 +799,46 @@ CQueryMutators::RunExtractAggregatesMutator(Node *node,
+			(Query *) node,
+			(MutatorWalkerFn) CQueryMutators::RunExtractAggregatesMutator,
+			context,
+			QTW_IGNORE_RT_SUBQUERIES  // flag -- do not mutate range table entries
```
This flag lets to ignore RTE_SUBQUERY only, but the comment says that all RTEs are not mutated

bimboterminator1 replied:
done

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -517,3 +517,50 @@ fetch backward all in c1;
+drop table if exists s;
+-- end_ignore
+create table t (i int, j int) distributed by (i);
+create table s (i int, j int) distributed by (i);
```
I suggest to put dropping and creating of the s table to the first usage of this table. First two tests do not use s. Dropping at the end should be in reverse order

bimboterminator1 replied:
done

bimboterminator1 committed:
Fix ORCA SubLink processing at query normalization
The ORCA optimizer could produce an invalid query plan or fallback to the legacy
optimizer due to incorrect processing of nested SubLinks (SubLink contains one
more SubLink or rtable subquery inside) during query normalization.

Firstly, the incorrect behaviour took place when a query had a nested
SubLink in its targetList, and SubLink lower levels contained attributes that
were referenced in the query's GROUP BY clause. ORCA could construct for
such queries an invalid plan, according to which some wrong and unexpected
attributes got into the nested part of SubLink's plan (the example is provided
in tests section). And the presence of those attributes could lead to incorrect
query execution or even segmentation fault.

Inside the NormalizeGroupByProjList function the context.m_lower_table_tlist,
which contains grouping attributes from target_list_copy, is created. And the
entries of target_list_copy, that are not grouping attributes, are mutated
inside the RunExtractAggregatesMutator in order to make their inner expressions
reference the corresponding entries from context.m_lower_table_tlist.
The invalid plan could be produced because of incorrect mapping between Var's
varattno (in target_list_copy) and some TargetEntry's resno
(in context.m_lower_table_tlist). This mapping is established by
RunExtractAggregatesMutator, where for the Var code branch the following
assignment takes place: var->varattno = found_tle->resno, where found_tle is the
corresponding entry from context.m_lower_table_tlist. However, the behaviour of
this function was erroneous for the case of nested SubLinks because it
mutated only the upper level of nested SubLink structure, and other subqueries
remained unmodified because RunExtractAggregatesMutator did not have a branch
for the Query, which was required for futher diving into lower subqueries. And
because inner Vars hadn't been mutated, they started reference wrong attributes.

This patch modifies RunExtractAggregatesMutator function by adding the Query
branch to make the mutator correctly step into nested SubLink queries. And the
call of MutateQueryOrExpressionTree is replaced with the call of mutator itself
in order to correctly modify m_current_query_level when stepping into lower
subqueries.

Secondly, the erroneous behaviour took place when nested SubLink was in
GROUP BY clause itself (see the test section). In this case nested SubLinks were
mutated incorrectly by RunGroupingColMutator.
RunGroupingColMutator had Query branch, however, when mutating the subquery
of the SubLink (at second query level), the m_current_query_level variable was
not increased respectively, because MutateQueryOrExpressionTree, which was
responsible for calling the RunGroupingColMutator on the lower subquery, did not
know anything about a context. Therefore, for the query like in the tests, wrong
value of m_current_query_level led to unnecessary modification of the Var's
varlevelsup value. And when Var's varlevelsup value had been unnecessary
modified, it started referencing different query level. This could lead to the
fallback or invalid plan construction at further planning stages.

This patch fixes erroneous behaviour of RunGroupingColMutator by replacing the
call of MutateQueryOrExpressionTree with the call of RunGroupingColMutator in
order to properly modify m_current_query_level when mutating lower level rtes.

Thirdly, similar issue could arise when query had correlated subquery in SubLink
under the AggRef.
While mutating the SubLink inside RunExtractAggregatesMutator, the Aggref branch
is executed for SubLink's query targetList, which mutates args of the AggRef.
And if a Var had varlevelsup value greater than the agglevelsup, i.e the Var
referenced a relation that is higher than the AggRef, the Var was not mutated
at all. The varlevelsup field was not modified and because of that, when AggRef
was pulled up to the zero query level (by appending it to the
context.m_lower_table_tlist), the Var, whose varlevelsup was higher than
AggRef's, and which remained unchanged, started referencing the wrong relation.
This could lead to the fallback or invalid plan construction when processing
attributes at further planning stages.

This patch makes the RunExtractAggregatesMutator modify the varlevelsup of the
Var relatively the AggRef level in order to preserve proper query level.

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -517,3 +517,52 @@ fetch backward all in c1;
+(select j from (select j) q2) q1
+from t group by i, j, q1;
+
+select j, 1 as c,
```
This query works correctly on adb-6.x-dev

andr-sokolov commented on src/test/regress/sql/subselect.sql:
```diff
@@ -517,3 +517,52 @@ fetch backward all in c1;
+explain (verbose, costs off)
+select (select max((select s.i from s where s.j = t.i))) from t;
+
+select (select max((select s.i from s where s.j = t.i))) from t;
```
This query works correctly on adb-6.x-dev

bimboterminator1 committed:
Formatting

bimboterminator1 committed:
Test description

bimboterminator1 commented:
Have been closed cause it has been decided to separate the solution into 3 single patches.
